les/servingqueue.go

```diff
@@ -159,27 +159,23 @@ func (sq *servingQueue) newTask(peer *clientPeer, maxTime uint64, priority int64
 // run tokens from the token channel and allow the corresponding tasks to run
 // without entering the priority queue.
 func (sq *servingQueue) threadController() {
+	defer sq.wg.Done()
 	for {
 		token := make(runToken)
 		select {
 		case best := <-sq.queueBestCh:
 			best.tokenCh <- token
 		case <-sq.stopThreadCh:
-			sq.wg.Done()
 			return
 		case <-sq.quit:
-			sq.wg.Done()
 			return
 		}
-		<-token
 		select {
 		case <-sq.stopThreadCh:
```

Comment:
I'm unsure if the `stopThreadCh` should take precedence over the `<-token` read. So originally, the code read `token`, and only afterwards would check if it's time to shut down. 
My _intended_ change was to also listen to `quit` events, but as an an un-intentional side-effect, my change maybe also modified the semantics for how thread management behaves. Previously it would fail to shut down a thread which was busy on a `token`, now it will just blindly kill it.

Reply:
But I really have no idea about the high level workings of this code, it's not exactly easy to understand just skimming through it

Reply:
Does not matter a lot, once the token has been granted to the task, it has no effect on the task any more whether we are waiting on the token to be closed. I'm not sure why I didn't do it like this in the first place.